ingest/boltcache.go

```diff
@@ -0,0 +1,867 @@
+
+package ingest
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+	"unsafe"
+
+	bolt "go.etcd.io/bbolt"
+
+	"github.com/gravwell/gravwell/v3/chancacher"
+	"github.com/gravwell/gravwell/v3/ingest/entry"
+)
+
+const (
+	defaultTickInterval = time.Second
+	defaultCacheSize    = 1024 * 1024 * 4   //4MB
+	defaultMaxCacheSize = 1024 * 1024 * 512 //512MB
+)
+
+var (
+	dbTimeout       time.Duration = 100 * time.Millisecond
+	dbMmapSize      int           = defaultCacheSize
+	dbOpenMode      os.FileMode   = 0660                        //user and group R/W but nothing for other
+	dbBucketName    []byte        = []byte(`ic`)                // this is the bucket that will hold entries
+	dbTagBucketName []byte        = []byte(`tagmap`)            // this bucket will hold the tag list
+	dbTagKey        []byte        = []byte(`__CACHE_TAG_KEY__`) // special tag to hold tag list
+
+	ErrActiveHotBlocks        = errors.New("There are active hotblocks, close pitched data")
+	ErrNoActiveDB             = errors.New("No active database")
+	ErrNoKey                  = errors.New("No key set on entry block")
+	ErrCacheAlreadyRunning    = errors.New("Cache already running")
+	ErrCacheNotRunning        = errors.New("Cache is not running")
+	ErrBucketMissing          = errors.New("Cache bucket is missing")
+	ErrCannotSyncWhileRunning = errors.New("Cannot sync while running")
+	ErrCannotPopWhileRunning  = errors.New("Cannot pop while running")
+	ErrInvalidKey             = errors.New("Key bytes are invalid")
+	ErrBoltLockFailed         = errors.New("Failed to acquire lock for ingest cache.  The file is locked by another process")
+)
+
+type IngestCacheConfig struct {
+	FileBackingLocation string
+	TickInterval        time.Duration
+	MemoryCacheSize     uint64
+	MaxCacheSize        uint64
+}
+
+type IngestCache struct {
+	mtx             *sync.Mutex
+	fileBacked      bool   //whether we are going to push to a file when there are no outputs available
+	storeLoc        string //location of boltDB
+	storedBlocks    int
+	count           uint64
+	cacheSize       uint64
+	storeSize       uint64
+	maxMemCacheSize uint64
+	maxCacheSize    uint64
+	hotBlocks       map[entry.EntryKey]*entry.EntryBlock
+	currKey         entry.EntryKey
+	currBlock       *entry.EntryBlock //just a pointer into the hotBlocks map value, NOT A COPY
+	db              *bolt.DB
+	err             error
+	running         bool
+	wg              sync.WaitGroup
+	stCh            chan bool
+}
+
+func boltTransition(c MuxerConfig) error {
+	// check if we have a bolt cache
+	// with regular files we'll just assume are bolt caches...
+	fi, err := os.Stat(c.CachePath)
+	if err != nil {
+		return nil
+	}
+	if fi.IsDir() {
+		return nil
+	}
+
+	// We probably do. Create a temporary chancacher, read the bolt cache
+	// into it, and then shuffle the files around.
+	tmpd, err := ioutil.TempDir("", "ingestCacheTransition")
```

Comment:
where is this tempdir firing off?  We want to make sure it lands on the same location as the existing cache in case anyone has a dedicated cache dir (I often do).

Reply:
The cache eventually gets back to the specified location but we can fire a temp file in the filepath.Base() of the provided dir.

Reply:
Updated.